### PR TITLE
MNT Add type annotations to pyodide_build/buildpkg.py

### PR DESCRIPTION
**Pull request by chigozienri**
Contributes to #718 
I'm not confident that I got all of these types right, so I'm pushing this for review before doing any of the other modules suggested in #718

### CODE REVIEW

File: pyodide_build/buildpkg.py [hunk 1]
```diff
@@ -103,7 +104,7 @@ def download_and_extract(buildpath, packagedir, pkg, args):
         raise ValueError("Incorrect source provided")
 
 
-def patch(path, srcpath, pkg, args):
+def patch(path: Path, srcpath: Path, pkg: Dict[str, Any], args: List(Any)):
```
Comment (rth) [inline pyodide_build/buildpkg.py:107]: I think `List(Any)` is syntactically not correct. It should be `List[Any]` or just `List` actually.

File: pyodide_build/buildpkg.py [hunk 2]
```diff
@@ -300,7 +301,7 @@ def make_parser(parser):
     return parser
 
 
-def main(args):
+def main(args: List(Any)):
```
Comment (rth) [inline pyodide_build/buildpkg.py:304]: I'm not sure this is right. After a quick glance I wasn't able to find the return type of `ArgumentParser.parse_args`. It's also fine to leave it untyped.

**Comment by chigozienri** [comment]
Results of running mypy:
```
pyodide_build/buildpkg.py:167: error: Dict entry 0 has incompatible type "str": "Path"; expected "str": "str"
pyodide_build/buildpkg.py:167: error: Dict entry 1 has incompatible type "str": "Path"; expected "str": "str"
pyodide_build/buildpkg.py:209: error: Missing return statement
Found 3 errors in 1 file (checked 9 source files)
```
Cast Path to str in `compile()`? (I can't work out type annotation is actually complaining here)
Add `return False` to the end of `needs_rebuild()`? Or remove `-> bool` from the function signature?

**Comment by rth** [comment]
> Cast Path to str in compile()? (I can't work out type annotation is actually complaining here)

It's complaining as `env` variable was defined earlier with a type `Dict[str, str]` and it keeps this type even when we redefine this later, and try to assign `Path` as value. If we create another variable `Dict[str, Path]` we get,
```
pyodide_build/buildpkg.py:168: error: Argument "env" to "run" has incompatible type "Dict[str, Path]"; expected "Union[Mapping[bytes, Union[bytes, str]], Mapping[str, Union[bytes, str]], None]"
```
So I think it's fine to explicitly cast to str when setting that dict. Environment variables have to be str anyway, and the conversion Path -> str was likely previously happening implicitly .


> Add return False to the end of needs_rebuild()? 

Yes, sounds good.

**Comment by rth** [comment]
I push a few fixed discussed above. Thanks for the PR @chigozienri !

**Comment by chigozienri** [comment]
Thanks for picking this up @rth , sorry I was slow on it!

**Pull request merged**